core/trie/proof.go

```diff
@@ -54,623 +55,586 @@ func (e *Edge) Len() uint8 {
+//
+//   - All elements proof: In this case the proof can be nil, but the range should be all the leaves in the trie.
+//   - One element proof: In this case, both left and right edge proofs are the same and the range should contain only one element.
+//   - Zero element proof: In this case a single edge proof is enough to verify. If there are still more elements, the proof is invalid.
+//
+// If the range proof is invalid, this function returns false and an error.
+// Otherwise, it will return a flag to indicate if there are more elements.
+//
+// TODO(weiihann): Given a binary leaf and a left-sibling first key, if the right sibling is removed, the proof would still be valid.
+// Conversely, given a binary leaf and a right-sibling last key, if the left sibling is removed, the proof would still be valid.
+// Range proof should not be valid for both of these cases, but currently is, which is an attack vector.
+// The problem probably lies in how we do root hash calculation.
+func VerifyRangeProof(root, first *felt.Felt, keys, values []*felt.Felt, proof *ProofNodeSet) (bool, error) { //nolint:funlen,gocyclo
+	// Ensure the number of keys and values are the same
+	if len(keys) != len(values) {
+		return false, fmt.Errorf("inconsistent proof data, keys: %d, values: %d", len(keys), len(values))
 	}
 
-	rootHash := leftPath[0].Hash(hash)
-
-	// Get duplicates and insert by one
-	i := 0
-	for i = 0; i < minLen; i++ {
-		leftNode := leftPath[i]
-		rightNode := rightPath[i]
+	// Ensure all keys are monotonic increasing and values contain no deletions
+	for i := 0; i < len(keys); i++ {
+		if i < len(keys)-1 && keys[i].Cmp(keys[i+1]) > 0 {
+			return false, errors.New("keys are not monotonic increasing")
+		}
 
-		if leftNode.Hash(hash).Equal(rightNode.Hash(hash)) {
-			merged = append(merged, leftNode)
-		} else {
-			break
+		if values[i] == nil || values[i].Equal(&felt.Zero) {
+			return false, errors.New("range contains deletion")
 		}
 	}
 
-	// Add rest of the nodes
-	merged = append(merged, leftPath[i:]...)
-	merged = append(merged, rightPath[i:]...)
-
-	return merged, rootHash, nil
-}
-
-// SplitProofPath splits the merged proof path into two paths (left and right), which were merged before
-// it first validates that the merged path is not circular, the split happens at most once and rootHash exists
-// then calls traverseNodes to split the path to left and right paths
-func SplitProofPath(mergedPath []ProofNode, rootHash *felt.Felt, hash hashFunc) ([]ProofNode, []ProofNode, error) {
-	commonPath := []ProofNode{}
-	leftPath := []ProofNode{}
-	rightPath := []ProofNode{}
-	nodeHashes := make(map[felt.Felt]ProofNode)
-
-	for _, node := range mergedPath {
-		nodeHash := node.Hash(hash)
-		_, nodeExists := nodeHashes[*nodeHash]
-
-		if nodeExists {
-			return leftPath, rightPath, errors.New("duplicate node in the merged path")
+	// Special case: no edge proof at all, given range is the whole leaf set in the trie
```

Comment:
```suggestion
	// Special case: no edge proof provided; the given range contains all leaves in the trie
```

The `set` suggests no-ordering